be/src/env/env_hdfs.cpp

```diff
@@ -2,137 +2,80 @@
-        }
-    }
-    _opened = true;
-    return Status::OK();
-}
-
-void HdfsRandomAccessFile::close() {
-    if (_opened) {
-        if (_fs && _file) {
-            hdfsCloseFile(_fs, _file);
-        }
-        _opened = false;
-    }
-}
-
-Status HdfsRandomAccessFile::_read_at(int64_t offset, char* data, size_t size, size_t* read_size) const {
-    if (_usePread) {
-        *read_size = size;
-        if (hdfsPreadFully(_fs, _file, offset, data, size) == -1) {
-            return Status::IOError(strings::Substitute("fail to hdfsPreadFully file, file=$0, error=$1", _file_name,
-                                                       get_hdfs_err_msg()));
-        }
-    } else {
-        auto cur_offset = hdfsTell(_fs, _file);
-        if (cur_offset == -1) {
-            return Status::IOError(
-                    strings::Substitute("fail to get offset, file=$0, error=$1", _file_name, get_hdfs_err_msg()));
-        }
-        if (cur_offset != offset) {
-            if (hdfsSeek(_fs, _file, offset)) {
-                return Status::IOError(strings::Substitute("fail to seek offset, file=$0, offset=$1, error=$2",
-                                                           _file_name, offset, get_hdfs_err_msg()));
-            }
-        }
-        size_t bytes_read = 0;
-        while (bytes_read < size) {
-            size_t to_read = size - bytes_read;
-            auto hdfs_res = hdfsRead(_fs, _file, data + bytes_read, to_read);
-            if (hdfs_res < 0) {
-                return Status::IOError(strings::Substitute("fail to hdfsRead file, file=$0, error=$1", _file_name,
-                                                           get_hdfs_err_msg()));
-            } else if (hdfs_res == 0) {
-                break;
-            }
-            bytes_read += hdfs_res;
-        }
-        *read_size = bytes_read;
-    }
-    return Status::OK();
+// class for remote read hdfs file
+// Now this is not thread-safe.
+class HdfsRandomAccessFile : public RandomAccessFile {
+public:
+    HdfsRandomAccessFile(hdfsFS fs, hdfsFile file, const std::string& file_name)
+            : _fs(fs), _file(file), _file_name(file_name), _file_size(0) {}
+
+    ~HdfsRandomAccessFile() override;
+
+    Status read(uint64_t offset, Slice* res) const override;
+    Status read_at(uint64_t offset, const Slice& res) const override;
+    Status readv_at(uint64_t offset, const Slice* res, size_t res_cnt) const override;
+    Status size(uint64_t* size) const override;
+    const std::string& file_name() const override { return _file_name; }
+    StatusOr<std::unique_ptr<NumericStatistics>> get_numeric_statistics() override;
+
+private:
+    hdfsFS _fs;
+    hdfsFile _file;
+    std::string _file_name;
+    mutable uint64_t _file_size;
+};
+
+HdfsRandomAccessFile::~HdfsRandomAccessFile() {
+    int r = hdfsCloseFile(_fs, _file);
+    PLOG_IF(ERROR, r != 0) << "close " << _file_name << " failed";
```

Comment:
what's this macro? `PREDICATE_LOG_IF`?

Reply:
> what's this macro? `PREDICATE_LOG_IF`?

FYI: https://github.com/google/glog#id21